File: objc/objcsearch/objcsearch/Regex.m
#import "Regex.h"

@implementation Regex

- (instancetype) initWithPattern:(NSString *)pattern {
    self = [super init];
    if (self) {
        self.pattern = pattern;
        NSError *error = NULL;
        self.expression = [NSRegularExpression
                           regularExpressionWithPattern:self.pattern
                           options:NSRegularExpressionDotMatchesLineSeparators
                           error:&error];
    }
    return self;
}

- (NSArray<NSTextCheckingResult*>*) matches:(NSString*) s {
    return [self.expression matchesInString:s
                                    options:0
                                      range:NSMakeRange(0, [s length])];
}

- (NSTextCheckingResult*) firstMatch:(NSString*) s {
    return [self.expression firstMatchInString:s
                                       options:0
                                         range:NSMakeRange(0, [s length])];
}

- (BOOL) test:(NSString *)s {
    return [self.expression numberOfMatchesInString:s
                                            options:0
                                              range:NSMakeRange(0, [s length])] > 0;
}

- (NSString *)description {
    return self.pattern;
}

@end

File: swift/swiftsearch/swiftsearch/Regex.swift
//
//  Regex.swift
//  swiftsearch
//
// from http://benscheirman.com/2014/06/regex-in-swift/ with modifications
//

import Foundation

class Regex {
    let expression: NSRegularExpression
    let pattern: String

    init(_ pattern: String) {
        self.pattern = pattern
        //var error: NSError?
        self.expression = try! NSRegularExpression(pattern: pattern,
            options: .dotMatchesLineSeparators)
    }

    func matches(_ s: String) -> [NSTextCheckingResult] {
        return self.expression.matches(in: s, options: [],
            range: NSMakeRange(0, s.lengthOfBytes(using: String.Encoding.utf8)))
    }

    func test(_ s: String) -> Bool {
        return matches(s).count > 0
    }
}
